File: nikukyu-web/src/services/account.service.ts
import {Account, AccountAuthorization, AccountLog, AccountOAuth} from "../data/account.ts"
import {get, list, remove, update} from "./service.ts"

type AccountParams = {
    nickname: string
    avatar: string
    background: string
    color: string
}

const getAccount = (accountId: string): Promise<Account> => get(`account/${accountId}`)
const updateAccount = (accountId: string, data: AccountParams): Promise<Account> =>
    update(`account/${accountId}`, data)

const listAccountAuthorization =
    (accountId: string, page: number, size: number): Promise<Array<AccountAuthorization>> =>
        list(`account/${accountId}/authorization?page=${page}&size=${size}`)
const getAccountAuthorization = (accountId: string, accountAuthorizationId: string): Promise<AccountAuthorization> =>
    get(`account/${accountId}/authorization/${accountAuthorizationId}`)
const deleteAccountAuthorization = (accountId: string, accountAuthorizationId: string): Promise<null> =>
    remove(`account/${accountId}/authorization/${accountAuthorizationId}`)

const listAccountLog = (accountId: string, page: number, size: number): Promise<Array<AccountLog>> =>
    list(`account/${accountId}/log?page=${page}&size=${size}`)
const getAccountLog = (accountId: string, logId: string): Promise<AccountLog> =>
    get(`account/${accountId}/log/${logId}`)

const listAccountOAuth = (accountId: string, page: number, size: number): Promise<Array<AccountOAuth>> =>
    list(`account/${accountId}/oauth?page=${page}&size=${size}`)
const getAccountOAuth = (accountId: string, oauthId: string): Promise<AccountOAuth> =>
    get(`account/${accountId}/oauth/${oauthId}`)
const deleteOAuth = (accountId: string, oauthId: string): Promise<null> =>
    remove(`account/${accountId}/oauth/${oauthId}`)

export {
    getAccount,
    updateAccount,
    listAccountAuthorization,
    getAccountAuthorization,
    deleteAccountAuthorization,
    listAccountLog,
    getAccountLog,
    listAccountOAuth,
    getAccountOAuth,
    deleteOAuth
}

File: nikukyu-server/nikukyu-server/src/main/java/io/hanbings/nikukyu/server/service/AccountService.java
package io.hanbings.nikukyu.server.service;

import io.hanbings.nikukyu.common.model.Account;
import io.hanbings.nikukyu.common.model.AccountAuthorization;
import io.hanbings.nikukyu.common.model.AccountLog;
import io.hanbings.nikukyu.common.model.AccountOAuth;
import io.hanbings.nikukyu.common.utils.RandomUtils;
import io.hanbings.nikukyu.common.utils.TimeUtils;
import io.hanbings.nikukyu.server.repository.AccountAuthorizationRepository;
import io.hanbings.nikukyu.server.repository.AccountLogRepository;
import io.hanbings.nikukyu.server.repository.AccountOAuthRepository;
import io.hanbings.nikukyu.server.repository.AccountRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Set;

@Service
@RequiredArgsConstructor
public class AccountService {
    static int MAX_PAGE_SIZE = 20;
    final AccountAuthorizationRepository accountAuthorizationRepository;
    final AccountLogRepository accountLogRepository;
    final AccountOAuthRepository accountOAuthRepository;
    final AccountRepository accountRepository;

    public Account createAccount(
            boolean verified,
            String id,
            String nick,
            String avatar,
            String background,
            String color,
            String email
    ) {
        return accountRepository.save(
                new Account(
                        RandomUtils.uuid(),
                        TimeUtils.getMilliUnixTime(),
                        verified,
                        id,
                        nick,
                        avatar,
                        background,
                        color,
                        email
                )
        );
    }

    public Account getAccount(String accountId) {
        return accountRepository.findAccountByAccountId(accountId);
    }

    public Account getAccountWithEmail(String email) {
        return accountRepository.findAccountByEmail(email);
    }

    public Account updateAccount(String accountId, String nickname, String avatar, String background, String color) {
        Account account = accountRepository.findAccountByAccountId(accountId);
        if (account == null) return null;

        Account updated = new Account(
                account.accountId(),
                account.created(),
                account.verified(),
                account.username(),
                nickname,
                avatar,
                account.email(),
                background,
                color
        );

        return accountRepository.save(updated);
    }

    public List<AccountAuthorization> getAccountAuthorizationList(String accountId, int page, int size) {
        Pageable pageable = PageRequest.of(
                page,
                Math.min(size, MAX_PAGE_SIZE)
        );
        Page<AccountAuthorization> accountAuthorizationPage = accountAuthorizationRepository.getAccountAuthorizationList(accountId, pageable);

        return accountAuthorizationPage.getContent();
    }

    public AccountAuthorization getAccountAuthorization(String accountId, String accountAuthorizationId) {
        return accountAuthorizationRepository.getAccountAuthorizationByCreatedByAndAccountAuthorizationId(accountId, accountAuthorizationId);
    }

    public AccountAuthorization createAccountAuthorization(
            String accountId,
            String provider,
            String openid
    ) {
        return accountAuthorizationRepository.save(
                new AccountAuthorization(
                        RandomUtils.uuid(),
                        TimeUtils.getMilliUnixTime(),
                        accountId,
                        provider,
                        openid
                )
        );
    }

    public AccountAuthorization getAccountAuthorization(String openId) {
        return accountAuthorizationRepository.getAccountAuthorizationByOpenid(openId);
    }

    public AccountAuthorization deleteAccountAuthorization(String accountId, String accountAuthorizationId) {
        return accountAuthorizationRepository.deleteAccountAuthorizationByAccountAuthorizationId(accountAuthorizationId);
    }

    public List<AccountLog> getAccountLogList(String accountId, int page, int size) {
        Pageable pageable = PageRequest.of(page, Math.min(size, MAX_PAGE_SIZE));
        Page<AccountLog> accountLogPage = accountLogRepository.getAccountLogList(accountId, pageable);

        return accountLogPage.getContent();
    }

    public AccountLog getAccountLog(String accountId, String accountLogId) {
        return accountLogRepository.getAccountLogByCreatedByAndAccountLogId(accountId, accountLogId);
    }

    public List<AccountOAuth> getAccountOAuthList(String accountId, int page, int size) {
        Pageable pageable = PageRequest.of(page, Math.min(size, MAX_PAGE_SIZE));
        Page<AccountOAuth> accountOAuthPage = accountOAuthRepository.getAccountOAuthList(accountId, pageable);

        return accountOAuthPage.getContent();
    }

    public AccountOAuth createAccountOAuth(String accountId, String oauthId, Set<String> access) {
        AccountOAuth accountOAuth = new AccountOAuth(
                RandomUtils.uuid(),
                TimeUtils.getMilliUnixTime(),
                accountId,
                oauthId,
                access
        );

        return accountOAuthRepository.save(accountOAuth);
    }

    public AccountOAuth getAccountOAuth(String accountId, String accountOAuthId) {
        return accountOAuthRepository.getAccountOAuthByCreatedByAndAccountOAuthId(accountId, accountOAuthId);
    }

    public AccountOAuth deleteAccountOAuth(String accountId, String accountOAuthId) {
        return accountOAuthRepository.deleteAccountOAuthByCreatedByAndAccountOAuthId(accountId, accountOAuthId);
    }
}
